File: solutions/120_max_path_sum.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// 又是看题解的一天
func maxPathSum(root *TreeNode) int {
	ans := -1 << 31
	var recursion func(root *TreeNode) int
	recursion = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		// MaxPathSum是从该节点至某叶子结点的最大路径和
		leftMaxPathSum := max(0, recursion(root.Left))
		rightMaxPathSum := max(0, recursion(root.Right))
		// 考虑 left + root + right 可能构成最大值
		tmp := leftMaxPathSum + rightMaxPathSum + root.Val
		if tmp > ans {
			ans = tmp
		}
		// 返回上层该root出发的最大路径和 若为负数 则不走比较好 即返回0
		rtv := max(leftMaxPathSum, rightMaxPathSum) + root.Val
		return max(0, rtv)
	}
	recursion(root)
	return ans
}

File: solutions/__120_max_path_sum.cpp
//
// Created by peanut on 2021/8/16.
//

#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

int ans = INT_MIN;

int maxGain(TreeNode *root) {
    if (root == nullptr)
        return 0;
    int left_max_gain = maxGain(root->left);
    int right_max_gain = maxGain(root->right);
    int tmp = left_max_gain + right_max_gain + root->val;
    if (tmp > ans)
        ans = tmp;
    int rtv = max(left_max_gain, right_max_gain) + root->val;
    return max(0, rtv);
}

int maxPathSum(TreeNode *root) {
    if (root == nullptr)
        return 0;
    maxGain(root);
    return ans;
}
